sink/datadog.go

```diff
@@ -92,29 +95,155 @@ func NewDatadog(monitorId string, opts, tags map[string]string, httpClient *http
 				return nil, err
 			}
 			d.tr = tr
+
 		case "metric-prefix":
 			if v == "" {
 				return nil, fmt.Errorf("datadog sink metric-prefix is empty string; value required when option is specified")
 			}
 			d.prefix = v
 
+		case "use-dogstatsd":
+			d.useDogStatsD = blip.Bool(v)
+
+		case "dogstatsd-host":
+			d.dogStatsDHost = v
+
+		case "dogstatsd-port":
+			d.dogStatsDPort = v
+
 		default:
 			return nil, fmt.Errorf("invalid option: %s", k)
 		}
 	}
 
-	if d.apiKeyAuth == "" {
-		return nil, fmt.Errorf("datadog sink required either api-key-auth or api-key-auth-file")
-	}
+	if d.useDogStatsD {
+		if d.dogStatsDHost == "" || d.dogStatsDPort == "" {
+			return nil, fmt.Errorf("dogStatsD host and port are required when DogStatsD is enabled")
+		}
+
+		client, err := statsd.New(fmt.Sprintf("%s:%s", d.dogStatsDHost, d.dogStatsDPort))
+		if err != nil {
+			return nil, err
+		}
+		d.dogStatsDClient = client
+	} else {
+		if d.apiKeyAuth == "" {
+			return nil, fmt.Errorf("datadog sink required either api-key-auth or api-key-auth-file")
+		}
 
-	if d.appKeyAuth == "" {
-		return nil, fmt.Errorf("datadog sink required either app-key-auth or app-key-auth-file")
+		if d.appKeyAuth == "" {
+			return nil, fmt.Errorf("datadog sink required either app-key-auth or app-key-auth-file")
+		}
+
+		c := datadog.NewConfiguration()
+		c.HTTPClient = httpClient
+		metricsApi := datadogV2.NewMetricsApi(datadog.NewAPIClient(c))
+		d.metricsApi = metricsApi
 	}
 
 	return d, nil
 }
 
 func (s *Datadog) Send(ctx context.Context, m *blip.Metrics) error {
+	if s.useDogStatsD {
+		return s.sendDogStatsD(ctx, m)
+	}
+	return s.sendApi(ctx, m)
+}
+
+func (s *Datadog) sendDogStatsD(ctx context.Context, m *blip.Metrics) error {
+	status.Monitor(s.monitorId, s.Name(), "sending metrics")
+
+	// On return, set monitor status for this sink
+	n := 0
+	defer func() {
+		status.Monitor(s.monitorId, s.Name(), "last sent %d metrics at %d", n, time.Now())
+	}()
+
+	// Pre-alloc DogStatsD data points
+	for _, metrics := range m.Values {
+		n += len(metrics)
+	}
+	if n == 0 {
+		return fmt.Errorf("no Blip metrics were collected")
+	}
+	n = 0
+
+	// Convert each Blip metric value to an DogStatsD data point
+	for domain := range m.Values { // each domain
+		metrics := m.Values[domain]
+		var name string
+
+	METRICS:
+		for i := range metrics { // each metric in this domain
+
+			// Set full metric name: translator (if any) else Blip standard,
```

Comment:
All this stuff in this loop is a big DRY violation with the other func that sends via API. So rather than duplicating code, let's add

```go
if s.dogstatsd {
   s.dogStatsDClient.Count(name, int64(metrics[i].Value), tags, 1)
   // ...
}
```

Right inline with the existing code.

Then update the docs to note that both can be used at same time, and dogstatsd is sent first, followed by API.

Reply:
I was aware of the duplication, but I was trying to avoid branching, while there's opportunity fro DRYing up, there's also differences (the main one being one sends inline in the loop, another uses batching) in they way they work that will introduce branching. the `send` method is already pretty long, inlining the other path will make it even more long and complex. I will have a look and see what I can come up with.